server/src/main/java/com/cloud/resource/RollingMaintenanceManagerImpl.java

```diff
@@ -666,11 +682,36 @@ private Pair<Boolean, String> performCapacityChecksBeforeHostInMaintenance(Host
         return new Pair<>(true, "OK");
     }
 
+    protected Ternary<Integer, Integer, Integer> getComputeResourcesCpuSpeedAndRamSize(VMInstanceVO runningVM) {
+        ServiceOfferingVO serviceOffering = serviceOfferingDao.findById(runningVM.getServiceOfferingId());
+        Integer cpu = serviceOffering.getCpu();
+        Integer speed = serviceOffering.getSpeed();
+        Integer ramSize = serviceOffering.getRamSize();
+        if (serviceOffering.isDynamic()) {
+            List<UserVmDetailVO> vmDetails = userVmDetailsDao.listDetails(runningVM.getId());
+            if (CollectionUtils.isNotEmpty(vmDetails)) {
+                for (UserVmDetailVO vmDetail : vmDetails) {
+                    if (vmDetail.getName() != null && vmDetail.getValue() != null) {
+                        if (cpu == null && VmDetailConstants.CPU_NUMBER.equals(vmDetail.getName())) {
+                            cpu = Integer.valueOf(vmDetail.getValue());
+                        } else if (speed == null && VmDetailConstants.CPU_SPEED.equals(vmDetail.getName())) {
+                            speed = Integer.valueOf(vmDetail.getValue());
+                        } else if (ramSize == null && VmDetailConstants.MEMORY.equals(vmDetail.getName())) {
+                            ramSize = Integer.valueOf(vmDetail.getValue());
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
```suggestion
        if (!serviceOffering.isDynamic()) {
            return new Ternary<>(cpu, speed, ramSize);
        }
        List<UserVmDetailVO> vmDetails = userVmDetailsDao.listDetails(runningVM.getId());
        if (CollectionUtils.isEmpty(vmDetails)) {
            return new Ternary<>(cpu, speed, ramSize);
        }
        for (UserVmDetailVO vmDetail : vmDetails) {
            if (vmDetail.getName() != null && vmDetail.getValue() != null) {
                if (cpu == null && VmDetailConstants.CPU_NUMBER.equals(vmDetail.getName())) {
                    cpu = Integer.valueOf(vmDetail.getValue());
                } else if (speed == null && VmDetailConstants.CPU_SPEED.equals(vmDetail.getName())) {
                    speed = Integer.valueOf(vmDetail.getValue());
                } else if (ramSize == null && VmDetailConstants.MEMORY.equals(vmDetail.getName())) {
                    ramSize = Integer.valueOf(vmDetail.getValue());
                }
            }
        }
```

Reply:
It's probably something we can deal in a separate PR? cc @sureshanaparti

Reply:
Its a very simple change for readability, do we really need another PR for it?

Reply:
I agree with @JoaoJandre on this one, inverting the `ifs` makes for a more readable code, and easier to understand its logic. Could even invert another if, like the suggestion below.

If these changes were on a code that this patch did not initially create, I would agree with @rohityadavcloud, but as it has been introduced with this patch, I think it is better to address this before merging.

```suggestion
        if (!serviceOffering.isDynamic()) {
            return new Ternary<>(cpu, speed, ramSize);
        }

        List<UserVmDetailVO> vmDetails = userVmDetailsDao.listDetails(runningVM.getId());
        if (!CollectionUtils.isNotEmpty(vmDetails)) {
            return new Ternary<>(cpu, speed, ramSize);
        }

        for (UserVmDetailVO vmDetail : vmDetails) {
            if (vmDetail.getName() == null || vmDetail.getValue() == null) {
                continue;
            }

            if (cpu == null && VmDetailConstants.CPU_NUMBER.equals(vmDetail.getName())) {
                cpu = Integer.valueOf(vmDetail.getValue());
            } else if (speed == null && VmDetailConstants.CPU_SPEED.equals(vmDetail.getName())) {
                speed = Integer.valueOf(vmDetail.getValue());
            } else if (ramSize == null && VmDetailConstants.MEMORY.equals(vmDetail.getName())) {
                ramSize = Integer.valueOf(vmDetail.getValue());
            }
        }
```

Reply:
@JoaoJandre @BryanMLima , I think this version is actually less readable as it contains multiple exit points. I prefer leaving this as is, maybe re-factoring out an extra method. The cyclic complexity of this method is high and the extra exits does not reduce it. (and yes, I think 7 is high and I am not counting the `== null &&`) 
@sureshanaparti I won't minus one if you apply this but I do not agree with this suggestion.

Reply:
@sureshanaparti how do you want to handle this? It is ready to merge otherwise it seems to me.

Reply:
will me check and update this @DaanHoogland